orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -5,58 +5,106 @@
 #include "core/session/inference_session.h"
 #include "core/session/environment.h"
 
+#include "orttraining/training_api/include/utils.h"
 #include "orttraining/training_api/include/optimizer.h"
 
 namespace onnxruntime {
 namespace training {
 namespace api {
 
-namespace {
-
-Status CreateOrtValueFromOrtValue(
-    const OrtValue& src_ort_value,
-    OrtValue& dest_ort_value,
-    onnxruntime::InferenceSession* sess) {
-  const Tensor& tensor = src_ort_value.Get<Tensor>();
-  AllocatorPtr allocator = sess->GetAllocator(tensor.Location());
-  const TensorShape& tensor_shape = tensor.Shape();
-  MLDataType element_type = tensor.DataType();
-
-  auto p_tensor = std::make_unique<Tensor>(element_type, tensor_shape, allocator);
-  dest_ort_value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(), DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
-
+Status Optimizer::GenerateMomentumNamedStates() {
+  auto& param_named_optimizer_states = optimizer_state_.param_named_optimizer_states;
+  auto& optim_sess_state = optim_sess_->GetSessionState();
+  for (auto& pair : named_parameters_) {
+    if (pair.second->RequiresGrad()) {
+      param_named_optimizer_states.insert({pair.first, ParameterOptimizerState()});
+      ParameterOptimizerState& cur_param_optimizer_states = param_named_optimizer_states[pair.first];
+      for (auto& state_name : utils::MOMENT_STATE_NAMES) {
+        OrtValue param_state;
+        ORT_ENFORCE(utils::OrtValueLike(optim_sess_state, pair.second->Data(), param_state).IsOK(), "Error generating moment state for ", pair.first);
+        cur_param_optimizer_states.momentum_named_states.insert({state_name, std::move(param_state)});
+      }
+    }
+  }
   return Status::OK();
 }
 
-}  // namespace
+// Constructs the ortvalue inputs to be fed to the graph
+// each step
+Status Optimizer::ConstructInputs() {
```

Comment:
hopefully we will make this function simpler after using new adam operator.

Reply:
yes we should be able to improve it, but not just after the new adam. We need some additional metadata in the checkpoint to avoid string matching